File: DotNetGotchas/VBDN/ReflectionToSerialize/Modified2/Serialization/DerivedCar.vb
'DerivedCar.vb

Imports System.Runtime.Serialization

<Serializable()> _
Public Class DerivedCar
    Inherits Car

    Private someValue As Integer

    Public Sub New(ByVal year As Integer, _
        ByVal anEngine As Engine, ByVal val As Integer)

        MyBase.New(year, anEngine)
        someValue = val
    End Sub

    Public Sub New( _
        ByVal info As SerializationInfo, _
        ByVal context As StreamingContext)

        MyBase.New(info, context)

        SerializationHelper.SetData( _
            GetType(DerivedCar), Me, info)
    End Sub

    Public Overrides Sub GetObjectData( _
        ByVal info As SerializationInfo, _
        ByVal context As StreamingContext)

        MyBase.GetObjectData(info, context)
        SerializationHelper.GetData(GetType(DerivedCar), Me, info)
    End Sub


    Public Overrides Function tostring() As String
        Return MyBase.ToString() & ":" & someValue
    End Function

End Class

File: DotNetGotchas/CSharp/ReflectionToSerialize/modified2/Serialization/DerivedCar.cs
//DerivedCar.cs
using System;
using System.Runtime.Serialization;

namespace Serialization
{
	[Serializable]
	public class DerivedCar : Car
	{
		private int someValue;

		public DerivedCar(int year, Engine anEngine, int val)
			: base(year, anEngine)
		{
			someValue = val;
		}

		public DerivedCar(SerializationInfo info, 
			StreamingContext context) 
			: base(info, context)
		{
			SerializationHelper.SetData(
				typeof(DerivedCar), this, info);
		}

		public override void GetObjectData(SerializationInfo info, 
			StreamingContext context)
		{
			base.GetObjectData(info, context);
			SerializationHelper.GetData(
				typeof(DerivedCar), this, info);
		}

		public override string ToString()
		{
			return base.ToString () + ":" + someValue;
		}

	}
}
